linux/library/src/text_input_model.cc

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <flutter_desktop_embedding/text_input_model.h>
+
+#include <iostream>
+
+#include <flutter_desktop_embedding/common/platform_protocol.h>
+
+static constexpr char kComposingBaseKey[] = "composingBase";
+
+static constexpr char kComposingExtentKey[] = "composingExtent";
+
+static constexpr char kSelectionAffinityKey[] = "selectionAffinity";
+static constexpr char kAffinityDownstream[] = "TextAffinity.downstream";
+
+static constexpr char kSelectionBaseKey[] = "selectionBase";
+static constexpr char kSelectionExtentKey[] = "selectionExtent";
+
+static constexpr char kSelectionIsDirectionalKey[] = "selectionIsDirectional";
+
+static constexpr char kTextKey[] = "text";
+
+static constexpr char kUpdateEditingStateMethod[] =
+    "TextInputClient.updateEditingState";
+
+namespace flutter_desktop_embedding {
+
+TextInputModel::TextInputModel(int client_id)
+    : rep_(""),
+      client_id_(client_id),
+      selection_base_(rep_.begin()),
+      selection_extent_(rep_.begin()) {}
+
+TextInputModel::~TextInputModel() {}
+
+bool TextInputModel::SetEditingState(size_t selection_base,
+                                     size_t selection_extent,
+                                     const std::string &text) {
+  if (selection_base > selection_extent) {
+    return false;
+  }
+  // Only checks extent since it is implicitly greater-than-or-equal-to base.
+  if (selection_extent > text.size()) {
+    return false;
+  }
+  rep_ = std::string(text);
+  selection_base_ = rep_.begin() + selection_base;
+  selection_extent_ = rep_.begin() + selection_extent;
+  return true;
+}
+
+void TextInputModel::DeleteSelected() {
+  selection_extent_ = rep_.erase(selection_base_, selection_extent_);
+  selection_base_ = selection_extent_;
```

Comment:
Per offline discussion let's change to:
  selection_base_ = ...
  selection_extent_ = selection_base_
since it seemed more confusing to me as written.

Reply:
Same with some other spots below.

Reply:
Fixed! :grinning: